### ISSUE TITLE
Count duplicates server side

### ISSUE THREAD

**Issue by filippobottega** [linq2db/linq2db#2063]
I need to check if a table column contains duplicates.
I just need true or false as result.

This is the sample code with 2 different query:

``` C#
using EtlDatalogic.Settings;
using LinqToDB;
using LinqToDB.Common;
using LinqToDB.Data;
using System;
using System.Linq;

namespace LinqToDb.Sample2
{
    class Program
    {
        static void Main(string[] args)
        {
            DataConnection.DefaultSettings = new MySettings();
            Configuration.Sql.GenerateFinalAliases = true;
            Configuration.Linq.CompareNullsAsValues = false;

            using var wipDB = new WipDB("Wip");

            using var tempTable = wipDB.CreateTempTable("TEMP_TABLE", new TempClass[] { new TempClass() { Value = "PIPPO" }, new TempClass() { Value = "PLUTO" }, new TempClass() { Value = "PLUTO" } });

            var query1 = from item in tempTable select item;
            var duplicatesQuery1 = query1
                .GroupBy(item => item.Value)
                .Where(group => group.Count() > 1)
                .Select(item => item.Key)
                .Take(1)
                .Any();

            Console.WriteLine(wipDB.LastQuery);

            var duplicatesCte1 = query1
                .GroupBy(item => item.Value)
                .Where(group => group.Count() > 1)
                .Select(item => item.Key).AsCte("Duplicates");

            var duplicatesQuery2 = (from item in duplicatesCte1 select item).Take(1).Any();

            Console.WriteLine(wipDB.LastQuery);
        }

    }

    public class TempClass
    {
        public string Value { get; set; }
    }

    public partial class WipDB : LinqToDB.Data.DataConnection
    {
        public WipDB()
        {
            InitDataContext();
            InitMappingSchema();
        }

        public WipDB(string configuration)
            : base(configuration)
        {
            InitDataContext();
            InitMappingSchema();
        }
        partial void InitDataContext();
        partial void InitMappingSchema();
    }

}
```

SQL of duplicatesQuery1  is:

``` SQL
SELECT
	CASE WHEN EXISTS(
		SELECT
			*
		FROM
			TEMP_TABLE item
		WHERE
			ROWNUM <= 1
		GROUP BY
			item.Value
		HAVING
			Count(*) > 1
	) THEN 1 ELSE 0 END as c1
FROM SYS.DUAL
```

Result is **false**!

SQL of duplicatesQuery2 is:

``` SQL
WITH Duplicates AS
(
	SELECT
		*
	FROM
		TEMP_TABLE item_1
	GROUP BY
		item_1.Value
	HAVING
		Count(*) > 1
)
SELECT
	CASE WHEN EXISTS(
		SELECT
			*
		FROM
			Duplicates item_2
		WHERE
			ROWNUM <= 1
	) THEN 1 ELSE 0 END as c1
FROM SYS.DUAL
```

Result is **true**!

The first result is wrong, the second is correct.

 I don't understand if it's a bug or if I have to take care about how to call Take() and GroupBy.

PS: second query doesn't work because Duplicates and AS are joined in DuplicatesAS. You have to edit the query manually to test it in Oracle.

**UPDATE**
I have rewritten the code as:

``` C#
...
            var query1 = from item in tempTable select item;
            var hasDuplicates1 = query1
                .GroupBy(item => item.Value)
                .Where(group => group.Count() > 1)
                .Select(item => item.Key)
                .Take(1)
                .Any();

            Console.WriteLine(wipDB.LastQuery);

            var hasDuplicates2 = query1
                .GroupBy(item => item.Value)
                .Where(group => group.Count() > 1)
                .Select(item => item.Key)
                .AsCte("Duplicates")
                .Take(1)
                .Any();

            Console.WriteLine(wipDB.LastQuery);
...
```

Now the second query is:

``` SQL
WITH Duplicates (Value)
AS
(
        SELECT
                item.Value as Value_1
        FROM
                TEMP_TABLE item
        GROUP BY
                item.Value
        HAVING
                Count(*) > 1
)
SELECT
        CASE WHEN EXISTS(
                SELECT
                        *
                FROM
                        Duplicates t1
                WHERE
                        ROWNUM <= 1
        ) THEN 1 ELSE 0 END as c1
FROM SYS.DUAL
```

Note than `WITH DuplicatesAS` becomes `WITH Duplicates (Value)` !
The emitted query is correct and works.
I think this is related to #2061 

### Environment details
linq2db version: 3.0.0-preview.1
Database Server: Oracle
Database Provider: Oracle.ManagedDataAccess.Core (3.0.0-preview.1)
Operating system: Windows 10
.NET Framework:.NET Core 3.1

**Comment by sdanyliv** [comment]
Looks like a bug in handling Take for Oracle. Thanks for reporting. In 3.0 we rewritten this logic and looks like we have missed this case.
Anyway you do not need Take for checking existence of duplicates. `Any` is enough.

**Comment by filippobottega** [comment]
Thank you @sdanyliv 
The only concern is about the comparison of these two queries:

``` SQL
WITH Duplicates (Value)
AS
(
        SELECT
                item.Value as Value_1
        FROM
                TEMP_TABLE item
        GROUP BY
                item.Value
        HAVING
                Count(*) > 1
)
SELECT
        CASE WHEN EXISTS(
                SELECT
                        *
                FROM
                        Duplicates t1
                WHERE
                        ROWNUM <= 1
        ) THEN 1 ELSE 0 END as c1
FROM SYS.DUAL
```

Plan: 
![2020-01-28 15_20_43-Oracle SQL Developer _ DATALOGIC_WIP_1_MES_WIPUSER](https://user-images.githubusercontent.com/2587684/73271929-c82c1280-41e1-11ea-819b-ae8a2bbed799.png)


and:

``` SQL
WITH Duplicates (Value)
AS
(
        SELECT
                item.Value as Value_1
        FROM
                TEMP_TABLE item
        GROUP BY
                item.Value
        HAVING
                Count(*) > 1
)
SELECT
        CASE WHEN EXISTS(
                SELECT
                        *
                FROM
                        Duplicates t1
        ) THEN 1 ELSE 0 END as c1
FROM SYS.DUAL
```

Plan: 
![2020-01-28 15_18_33-Oracle SQL Developer _ DATALOGIC_WIP_1_MES_WIPUSER](https://user-images.githubusercontent.com/2587684/73271813-9450ed00-41e1-11ea-9952-4c1bf26dfe30.png)


I don't know which execution time is better in general and not only in this specific case. I think that the first is better.

**Comment by sdanyliv** [comment]
Why it is better? Optimizer should look only for first record in EXISTS operator. And it should be automatic process.

**Comment by filippobottega** [comment]
I don't know but the time is better in the first case, (0,186 compared to 0,25) and I don't know how EXISTS operator works. Anyway I'll check if there is any concerning documentation.

**Comment by MaceWindu** [comment]
Closed by #2065
